File: materials/plastic/plastic_interface.py
import numpy as np

from src.fem.materials._material import Material
import src.fem.materials.plastic.plastic as plastic
import src.base.consts as consts
from src.base.tensor import iso, dev
from src.base.errors import WasatchError

class Plastic(Material):
    name = "plastic"
    mid = 3
    def __init__(self):
        """Instantiate the Plastic material

        """
        super(Plastic, self).__init__()
        self.register_parameters({"K": 0, "MU": 1, "A1": 2, "A4": 3, "DEVEL": 4})

    def setup(self, pdict):
        """Set up the Elastic material

        Parameters
        ----------
        pdict : dict
            Parameter dictionary

        """
        self.parse_input_parameters(pdict)
        plastic.plastic_check(self._params)

    def initialize_state(self, *args, **kwargs):
        nxtra, namea, keya, xtra = plastic.plastic_request_xtra(self._params)
        self.register_variables(nxtra, namea, keya, mig=True)
        self.xtra = xtra
        pass

    def update_state(self, dt, d, stress, xtra):
        """Compute updated stress given strain increment

        Parameters
        ----------
        dt : float
            Time step

        d : array_like
            Deformation rate

        stress : array_like
            Stress at beginning of step

        xtra : array_like
            Extra variables

        Returns
        -------
        S : array_like
            Updated stress

        xtra : array_like
            Updated extra variables

        """
        plastic.plastic_update_state(dt, self._params, d, stress, xtra)
        return stress, xtra

    def stiffness(self, dt, d, stress, xtra):
        """Return the constant stiffness
        dt : float
            time step

        d : array_like
            Deformation rate

        stress : array_like
            Stress at beginning of step

        """
        J = plastic.plastic_stiff(dt, self._params, d, stress, xtra)
        return J

File: materials/plastic/plastic_interface.f90
subroutine plastic_check(ui)
  use plastic, only: pchk
  implicit none
  integer, parameter :: rk=selected_real_kind(14)
  integer, parameter :: nui=5
  real(kind=rk), intent(inout) :: ui(nui)
  call pchk(ui)
  return
end subroutine plastic_check
subroutine plastic_request_xtra(ui, nxtra, namea, keya, xtra)
  use plastic, only: prxv
  implicit none
  integer, parameter :: rk=selected_real_kind(14)
  integer, parameter :: nui=5, nx=2
  real(kind=rk), intent(in) :: ui(nui)
  integer, intent(out) :: nxtra
  character(1), intent(out) :: namea(nx*50), keya(nx*10)
  real(kind=rk), intent(out) :: xtra(nx)
  call prxv(ui, nxtra, namea, keya, xtra)
  return
end subroutine plastic_request_xtra
subroutine plastic_update_state(nx, dt, ui, d, stress, xtra)
  use plastic, only: pcalc
  integer, parameter :: rk=selected_real_kind(14)
  integer, parameter :: nui=5
  integer, intent(in) :: nx
  real(kind=rk), intent(in) :: dt, ui(nui), d(6)
  real(kind=rk), intent(inout) :: stress(6), xtra(nx)
  call pcalc(nx, dt, ui, d, stress, xtra)
  return
end subroutine plastic_update_state
subroutine plastic_stiff(nx, dt, ui, d, stress, xtra, J)
  use plastic, only: pstiff
  integer, parameter :: rk=selected_real_kind(14)
  integer, parameter :: nui=5
  integer, intent(in) :: nx
  real(kind=rk), intent(in) :: dt, ui(nui), d(6)
  real(kind=rk), intent(in) :: stress(6), xtra(nx)
  real(kind=rk), intent(out) :: J(6, 6)
  call pstiff(nx, dt, ui, d, stress, xtra, J)
  return
end subroutine plastic_stiff
